File: C++/BInary Trees/DiameterOfBT.cpp
#include <iostream>
#include <bits/stdc++.h>
// #include <queue>

using namespace std;

class Node
{
public:
    int data;
    Node *left;
    Node *right;

    Node(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

int Height(Node *root)
{
    if (root == nullptr)
        return 0;
    int left = Height(root->left);
    int right = Height(root->right);
    int ans = max(left, right) + 1;
    return ans;
}

int Diameter(Node* root){
    if(root == nullptr)
        return 0;
    int leftDiameter = Diameter(root->left);
    int rightDiameter = Diameter(root->right);
    int totalDiameter = Height(root->left) + Height(root->right) + 1;

    return max({leftDiameter, rightDiameter, totalDiameter});

}

pair<int, int>DiameterOptimized(Node* root){

    if(root == nullptr)
        return {0, 0};
        
    pair<int, int> left = DiameterOptimized(root->left);
    pair<int, int> right = DiameterOptimized(root->right);

    int leftDiameter = left.first;
    int rightDiameter = right.first;
    int totalDiameter = left.second + right.second + 1;

    pair<int, int> ans;
    ans.first = max({leftDiameter, rightDiameter, totalDiameter});
    ans.second = max(left.second, right.second) + 1;

    return ans;
}

void deleteTree(Node *root)
{
    if (root == nullptr)
        return;

    deleteTree(root->left);
    deleteTree(root->right);
    delete root; // Delete the current node AFTER deleting children
}

int main()
{
    Node *root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->right = new Node(6);
    root->left->right->left = new Node(7);
    root->right->right->left = new Node(8);
    root->right->right->left->left = new Node(9);

    // int diameter = Diameter(root);
    int diameter = DiameterOptimized(root).first;
    cout << "Diameter of BT is: " << diameter << endl;

    // ---------------------------

    deleteTree(root);

    return 0;
}

File: Python/Binary Trees/DiameterOfBT.py

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def DiameterOptimized(root):
    if root is None:
        return (0, 0)
    left = DiameterOptimized(root.left)
    right = DiameterOptimized(root.right)

    diameter = max(left[0], right[0], left[1] + right[1] + 1)
    height = max(left[1], right[1]) + 1

    return diameter, height


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    root.right.right = Node(6)
    root.left.right.left = Node(7)
    root.right.right.left = Node(8)
    root.right.right.left.left = Node(9)

    diameter, height = DiameterOptimized(root)
    print('Diameter of BT is: ', diameter)
